java/src/main/native/src/RmmJni.cpp

```diff
@@ -75,48 +75,32 @@ class tracking_resource_adaptor final : public base_tracking_resource_adaptor {
   Upstream *get_wrapped_resource() { return resource; }
 
   std::size_t get_total_allocated() override {
-    std::lock_guard<std::mutex> lock(size_map_mutex);
-    return total_allocated;
+    return total_allocated.load();
   }
 
 private:
   Upstream *const resource;
   std::size_t const size_align;
-  std::size_t total_allocated{0};
-
-  // map and associated lock to track memory sizes by address
-  // TODO: This should be removed when rmm::alloc and rmm::free are removed and the size parameter
-  //       for do_deallocate can be trusted. If map and mutex are removed then total_allocated
-  //       should be updated to be atomic.
-  std::unordered_map<void *, std::size_t> size_map{};
-  std::mutex size_map_mutex{};
+  std::atomic_size_t total_allocated{0};
 
   void *do_allocate(std::size_t num_bytes, cudaStream_t stream) override {
     // adjust size of allocation based on specified size alignment
     num_bytes = (num_bytes + size_align - 1) / size_align * size_align;
 
-    std::lock_guard<std::mutex> lock(size_map_mutex);
-
     auto result = resource->allocate(num_bytes, stream);
     if (result) {
       total_allocated += num_bytes;
-      size_map[result] = num_bytes;
     }
     return result;
   }
 
   void do_deallocate(void *p, std::size_t size, cudaStream_t stream) override {
-    std::lock_guard<std::mutex> lock(size_map_mutex);
-
+    size = (size + size_align - 1) / size_align * size_align;
+    
     resource->deallocate(p, size, stream);
 
     if (p) {
```

Comment:
Do we need to check for a nullptr here?  We're no longer trying to use the pointer to lookup a map and arguably only care about the reported size.

Reply:
I think it's better to do a check on it, if the caller provides a nullptr but with size > 0, nothing is freed, so total_allocated should not change